btcec/schnorr/musig2/sign.go

```diff
@@ -0,0 +1,418 @@
+	var (
+		nonceMsgBuf  bytes.Buffer
+		nonceBlinder btcec.ModNScalar
+	)
+	nonceMsgBuf.Write(combinedNonce[:])
+	nonceMsgBuf.Write(schnorr.SerializePubKey(combinedKey))
+	nonceMsgBuf.Write(msg[:])
+	nonceBlindHash := chainhash.TaggedHash(
+		NonceBlindTag, nonceMsgBuf.Bytes(),
+	)
+	nonceBlinder.SetByteSlice(nonceBlindHash[:])
+
+	var nonce, r1J, r2J btcec.JacobianPoint
+	r1.AsJacobian(&r1J)
+	r2.AsJacobian(&r2J)
+
+	// With our nonce blinding value, we'll now combine both the public
+	// nonces, using the blinding factor to tweak the second nonce:
+	//  * R = R_1 + b*R_2
+	btcec.ScalarMultNonConst(&nonceBlinder, &r2J, &r2J)
+	btcec.AddNonConst(&r1J, &r2J, &nonce)
+
+	// If the combined nonce it eh point at infinity, then we'll bail out.
+	if nonce == infinityPoint {
+		return nil, ErrNoncePointAtInfinity
+	}
+
+	// Next we'll parse out our two secret nonces, which we'll be using in
+	// the core signing process below.
+	//
+	// TODO(roasbeef): no > n check actually exists here since mod n
+	// scalar, need to check overflow value instead?
+	var k1, k2 btcec.ModNScalar
+	k1.SetByteSlice(secNonce[:btcec.PrivKeyBytesLen])
+	k2.SetByteSlice(secNonce[btcec.PrivKeyBytesLen:])
+
+	nonce.ToAffine()
+
+	nonceKey := btcec.NewPublicKey(&nonce.X, &nonce.Y)
+
+	// If the nonce R has an odd y coordinate, then we'll negate both our
+	// secret nonces.
+	if nonce.Y.IsOdd() {
+		k1.Negate()
+		k2.Negate()
+	}
+
+	privKeyScalar := privKey.Key
+	if privKeyScalar.IsZero() {
+		return nil, ErrPrivKeyZero
+	}
+
+	// If the y coordinate of the public key is odd xor the y coordinate of
+	// the combined public key is odd, then we'll negate the private key.
+	pubKey := privKey.PubKey()
+	pubKeyYIsOdd := func() bool {
+		pubKeyBytes := pubKey.SerializeCompressed()
+		return pubKeyBytes[0] == secp.PubKeyFormatCompressedOdd
+	}()
+	combinedKeyYIsOdd := func() bool {
+		combinedKeyBytes := combinedKey.SerializeCompressed()
+		return combinedKeyBytes[0] == secp.PubKeyFormatCompressedOdd
+	}()
+	if pubKeyYIsOdd != combinedKeyYIsOdd {
+		privKeyScalar.Negate()
+	}
+
+	// Next we'll create the challenge hash that commits to the combined
+	// nonce, combined public key and also the message: * e =
+	// H(tag=ChallengeHashTag, R || Q || m) mod n
+	var challengeMsg bytes.Buffer
+	challengeMsg.Write(schnorr.SerializePubKey(nonceKey))
+	challengeMsg.Write(schnorr.SerializePubKey(combinedKey))
+	challengeMsg.Write(msg[:])
+	challengeBytes := chainhash.TaggedHash(
+		ChallengeHashTag, challengeMsg.Bytes(),
+	)
+	var e btcec.ModNScalar
+	e.SetByteSlice(challengeBytes[:])
+
+	// Next, we'll compute mu, our aggregation coefficient for the key that
+	// we're signing with.
+	mu := aggregationCoefficient(pubKeys, pubKey)
+
+	// With mu constructed, we can finally generate our partial signature
+	// as: s = (k1_1 + b*k_2 + e*mu*d) mod n.
+	s := new(btcec.ModNScalar)
+	s.Add(&k1).Add(k2.Mul(&nonceBlinder)).Add(e.Mul(mu).Mul(&privKeyScalar))
+
+	sig := NewPartialSignature(s, nonceKey)
```

Comment:
I think we should overwrite the `secnonce` with all zeros here (might need to change it to a pointer type to have an effect) to avoid signing with the same nonce again. Could add a check at the beginning if it is all zeros.